beginend.el

```diff
@@ -61,107 +62,165 @@
   "Evaluate &BODY and goto real beginning if that did not change point."
   (let ((tempvar (make-symbol "old-position")))
     `(let ((,tempvar (point)))
+       (goto-char (point-min))
        ,@body
        (when (equal ,tempvar (point))
-         (call-interactively #'beginning-of-buffer)))))
+         (goto-char (point-min))))))
```

Comment:
Hm, I think calling `beginning-of-buffer` might not be a bad idea.  Or at least, we should probably update marks accordingly.  That is, before calling our "smart" function we should push mark from where we jumped as that is what `beginning-of-buffer` does.